src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -235,35 +250,32 @@ public override string ToString()
 
             var vsb = new ValueStringBuilder(stackalloc char[512]);
 
-            if (_headerStore is Dictionary<HeaderDescriptor, object> headerStore)
+            foreach (HeaderEntry entry in GetEntries())
```

Comment:
This (and other calls to GetEntries) will force an array allocation if we are in dictionary mode, right? Seems like we could avoid that without too much trouble.

Reply:
You mean by caching the array or by implementing a custom enumerator?
I tried to optimize for the common case, even if it made the dictionary edge-case slightly more expensive.

Reply:
I was thinking more like a custom enumerator. That said, I can see how this could potentially add cost to the common case. That's unfortunate. Seems like we should be able to handle both cases efficiently and without allocating, but it's not clear to me how to do this without significantly complicating the code. Hmmm.